awx/main/scheduler/task_manager.py

```diff
@@ -519,20 +519,27 @@ def process_pending_tasks(self, pending_tasks):
             if not found_acceptable_queue:
                 logger.debug("{} couldn't be scheduled on graph, waiting for next cycle".format(task.log_format))
 
-    def timeout_approval_node(self):
-        workflow_approvals = WorkflowApproval.objects.filter(status='pending').prefetch_related('workflow_approval_template')
+    def timeout_approval_node(self): # Add websocket stuff for when it transitions to "timed out" (maybe a websocket_emit_status() call)
+        workflow_approvals = WorkflowApproval.objects.filter(status='pending')
         now = tz_now()
         for task in workflow_approvals:
-            # TODO: copy the timeout to the job itself at launch time, not the template
-            approval_timeout_seconds = timedelta(seconds=task.workflow_approval_template.timeout)
-            if task.workflow_approval_template.timeout == 0:
+            # TODO: copy the timeout to the job itself at launch time, not the template <---- Started to implement steps to do this, but unsure...
+            approval_timeout_seconds = timedelta(seconds=task.timeout)
+            if task.timeout == 0:
                 continue
             if (now - task.created) >= approval_timeout_seconds:
-                logger.info("The approval node {} ({}) has expired after {} seconds.".format(task.name, task.id, task.workflow_approval_template.timeout))
+                logger.info("The approval node {} ({}) has expired after {} seconds.".format(task.name, task.pk, task.timeout))
                 task.status = 'failed'
                 task.job_explanation = _("This approval node has timed out.")
                 task.save(update_fields=['status', 'job_explanation'])
 
+    # &&&&&& Placeholder for websocket support
+    # def detect_pending_approval(self):
+    #     workflow_approvals = WorkflowApproval.objects.filter(status='pending').prefetch_related('workflow_approval_template')
```

Comment:
We've already got an existing mechanism for broadcasting job status changes.  We may not _actually_ have to add new explicit code to make this work (we might get it for free) cc @mabashian

Reply:
I'd expect this line of code to emit a `pending` notification for these approvals:

https://github.com/ansible/awx/blob/devel/awx/main/scheduler/task_manager.py#L285

Reply:
I don't know that we care very much about web socket status changes for approval/reject (assuming our main use case here is for the UI to notice *new* pending notifications, and alert the user).

Reply:
I think the _hardest_ part of the websocket implementation here isn't actually on the API.  In fact, because of the fact that we based this on the UJT -> UJ paradigm, I _think_ we get pending status notifications mostly for free.

The _hard part_ is the that the UI will have to listen for _these_ notifications on *every* page of the UI if it wants to actually update the global notification list automatically.

Reply:
We care about websocket status changes for the mundane task of updating the status shown of an approval node inside a workflow job in the UI. I agree with your sentiment that we may or may not require any extra work for that, we just have to launch it and watch.

> The hard part is that the UI will have to listen for these notifications on every page of the UI if it wants to actually update the global notification list automatically.

This gets to the crux of the hard decision which has not yet been made. For this global listening, will the UI be listening to the group jobs-status_changed?

https://github.com/ansible/awx/blob/5d6916f69ee1ce39676384514fc0cb79cbe3bdaa/awx/main/models/unified_jobs.py#L1192

That should be already hooked up, and if extra work is needed, it should be minimal.

But the larger problem with relying on the job status channel group is that the UI just doesn't work that way. It is frequently subscribing and unsubscribing from it. Perhaps the UI could re-design that... but that's a big jump.

Otherwise, the 🔔 demands a new channel group. Say it's called "workflow_approvals". From the moment the user is logged in, the UI would subscribe to that channel and remain subscribed.

Reply:
Yep, we can be flexible here.  I'd mostly like to hear @mabashian's thoughts on what he needs to make this happen.